### PR TITLE
Refactor/master no doc

### PR DESCRIPTION
**Pull request by daviswahl**
Full  reimplementation of the main Query interface, along with documentation and specs. Should lay the groundwork for other contributors, if they're so inclined.

I will probably be finishing up the project/sort expressions over the next few days. Once that's done, it will be ready to start sharing around. The DSL/sugaring won't be done but I think the expression composition syntax is close to being locked down, which is what matters.

Other notes
-- Switched from minitest to rspec. Minitest wasn't doing it for me.
-- Generate docs with `yard`

### CODE REVIEW

File: Rakefile [hunk 1]
```diff
@@ -1,14 +1,13 @@
...
+  RuboCop::RakeTask.new
+  RSpec::Core::RakeTask.new(:spec)
 
-Rake::TestTask.new do |t|
-  t.verbose = true
-  t.libs    = ['lib']
-  t.libs.push 'test'
-  t.test_files = FileList['test/**/*_test.rb']
+  task default: :spec
+rescue LoadError
```
Comment (jturel) [inline Rakefile]: Can this be removed?

**Comment by jturel** [comment]
You've set all the tests to be skipped!

File: Rakefile (hunk 1 above)
Comment (daviswahl) [inline Rakefile]: Yeah, it's just the recommended practice from the RSpec devs (it allows rake to be used in non development environments). We don't actually use rake outside of development so I don't think it's particularly important to keep around.

**Comment by daviswahl** [comment]
The old tests are disabled because they're not currently passing. They will be passing before this gets merged into master. 

The point of merging this into its own branch is so that you can keep track of the work while it's in progress, so you don't get sidelined by a 6,000 line pull request. But if you'd prefer that I just keep working on it until it's ready, that's fine too.

**Comment by jturel** [comment]
I'm confused now. Is this a representation of the work that was demo'd at the dev meeting or has it evolved into this diff ?

**Pull request closed without merging**